File: PolyBridge.Sample/Assets/Scripts/PaymentService.cs
using System.Threading.Tasks;
using Cysharp.Threading.Tasks;
using PolyBridge.Core.Attributes;
using PolyBridge.Sandbox;

[NativeService("com.polybridge.sample.payment.PaymentService",
    BridgeType = typeof(PaymentServiceBridge))]
[Sandbox("Payment Service")]
public partial class PaymentService
{
    // 동기 — 반환값
    [NativeMethod("isInitialized")]
    [SandboxMethod("Is Initialized")]
    public partial bool IsInitialized();

    [NativeMethod("getServiceName")]
    [SandboxMethod("Get Service Name")]
    public partial string GetServiceName();

    // 비동기 — UniTask void
    [NativeMethod("initialize")]
    [SandboxMethod("Initialize")]
    public partial UniTask InitializeAsync();

    // 비동기 — UniTask 복합 타입 반환
    [NativeMethod("purchase")]
    [SandboxMethod("Purchase")]
    public partial UniTask<PaymentResult> PurchaseAsync(string productId, int amount);

    // 비동기 — string 반환
    [NativeMethod("getReceipt")]
    [SandboxMethod("Get Receipt")]
    public partial Task<string> GetReceiptAsync(string transactionId);

    // Editor Mock
    [MockReturn(nameof(IsInitialized))]
    internal bool MockReturnIsInitialized() => true;

    [MockReturn(nameof(GetServiceName))]
    internal string MockReturnGetServiceName() => "PaymentService-mock";

    [MockReturn(nameof(InitializeAsync))]
    internal UniTask MockReturnInitializeAsync() => UniTask.CompletedTask;

    [MockReturn(nameof(PurchaseAsync))]
    internal UniTask<PaymentResult> MockReturnPurchaseAsync() => UniTask.FromResult(new PaymentResult
    {
        transactionId = "mock_txn_001",
        productId = "item_001",
        amount = 1000,
        currency = "KRW",
        status = "mock_completed"
    });

    [MockReturn(nameof(GetReceiptAsync))]
    internal Task<string> MockReturnGetReceiptAsync() => Task.FromResult("{\"transactionId\":\"mock_txn\",\"receipt\":\"mock_receipt\"}");
}

File: PolyBridge.Sample.Android/polybridge-sample/src/main/java/com/polybridge/sample/payment/PaymentService.kt
package com.polybridge.sample.payment

import org.json.JSONObject
import kotlin.concurrent.thread

class PaymentService {

    private var initialized = false
    private var eventListener: IPaymentBridge? = null

    fun setEventListener(listener: IPaymentBridge) {
        eventListener = listener
    }

    fun removeEventListener(listener: IPaymentBridge) {
        if (eventListener === listener) eventListener = null
    }

    // 동기 — 반환값 없음
    fun isInitialized(): Boolean = initialized

    // 동기 — 반환값
    fun getServiceName(): String = "PaymentService-v1.0"

    // 비동기 — void
    fun initialize(callback: IPaymentBridge) {
        thread {
            Thread.sleep(500)
            initialized = true
            eventListener?.onPaymentStateChanged("initialized")
            callback.onSuccess("initialized")
        }
    }

    // 비동기 — 반환값 (복합 타입)
    fun purchase(productId: String, amount: Int, callback: IPaymentBridge) {
        if (!initialized) {
            callback.onError("PaymentService not initialized")
            return
        }

        eventListener?.onPaymentStateChanged("processing")

        thread {
            Thread.sleep(1000)

            val result = PaymentResult(
                transactionId = "txn_${System.currentTimeMillis()}",
                productId = productId,
                amount = amount,
                currency = "KRW",
                status = "completed"
            )

            val json = JSONObject().apply {
                put("transactionId", result.transactionId)
                put("productId", result.productId)
                put("amount", result.amount)
                put("currency", result.currency)
                put("status", result.status)
            }

            eventListener?.onPaymentStateChanged("completed")
            eventListener?.onReceiptReady(result.transactionId, json.toString())
            callback.onSuccess(json.toString())
        }
    }

    // 비동기 — 반환값 (primitive string)
    fun getReceipt(transactionId: String, callback: IPaymentBridge) {
        if (!initialized) {
            callback.onError("PaymentService not initialized")
            return
        }

        thread {
            Thread.sleep(300)
            callback.onSuccess("{\"transactionId\":\"$transactionId\",\"receipt\":\"receipt_${System.currentTimeMillis()}\"}")
        }
    }
}
